95.md

```diff
@@ -0,0 +1,91 @@
+an https route at `/nip95(/<sha256-file-hash>)` for handling files as described below.
+They may announce support with [NIP-11](11.md) `supported_nips` field.
+Other ways of advertising NIP-95 support may be described later.
+
+## Upload
+
+A file can be uploaded one at a time to `https://relay.domain/nip95` as `multipart/form-data` content type using `POST` method with the file object set to the `file` form data field.
+
+Clients must add a [NIP-98](98.md) `Authorization` header with the encoded `payload` tag set to the base64-encoded 256-bit SHA-256 hash of the file (not the hash of the whole request body).
+If using an html form, use an `Authorization` form data field instead.
+
+Others custom form data fields may be added depending on specific relay support.
+
+The `filename` embedded in the file may not be honored by the relay, which can internally store just the filename as the SHA-256 hash value ignoring extra metadata.
+
+Relays must reject the upload if the `payload` tag value contains an already used SHA-256 hash or
+it isn't the same of the received file.
+
+The upload response is a json as follows:
+
+```js
+{
+  nip95: {
+    // SHA-256 hash. Empty string if unsuccessful
+    x: "719171db19525d9d08dd69cb716a18158a249b7b3b3ec4bbdec5698dca104b7b"
+  },
+  errors: {
+    // Empty array if successful
+    nip95: ['error 1', 'error 2']
+  }
+}
+```
+
+Clients may upload the same file to one or many relays.
+After successful upload, the client may optionally generate and send to any set of relays an [NIP-94](94.md) event with atleast these tags:
+
+```js
+[
+  ["url", "https://relay.domain/nip95/719171db19525d9d08dd69cb716a18158a249b7b3b3ec4bbdec5698dca104b7b"],
+  ["x", "719171db19525d9d08dd69cb716a18158a249b7b3b3ec4bbdec5698dca104b7b"]
+]
+```
+
+If the file upload was made to more than one relay, `r` query params must be added to the url string, containing each of the relays used in successful upload operations. For example:
+
+```js
+[
+  ["url", "https://relay.domain/nip95/<sha256-file-hash>?r=relay.domain2&r=relay.domain3&r=relay.domain4"],
+  ["x", <the-hash>]
+]
+```
+
+Alternativelly, instead of using NIP-94, one can share or embbed on a note just the above url, optionally with added [NIP-54](54.md) inline metadata.
+
+File compression is optionally done at the client, **never on relay**, before uploading.
+This is done to assure the same file hash result, no matter to how many relays one may upload.
+
+## Download
+
+Relays must make available the route path `/nip95/<sha256-file-hash>` with `GET` method for file download.
+
+Clients should send a `GET` request to the relay url in the format `https://relay.domain/nip95/<sha256-file-hash>?r=relay.domain2&r=relay.domain3`.
+
+If the relay doesn't have the file anymore, it should issue a 302 http redirect to the next relay included as `r` query param. For example, issue a redirect to `https://relay.domain2/nip95/<sha256-file-hash>?r=relay.domain3`
+
```

Comment:
This could result in an infinite redirect loop

Reply:
Yes. I will add something like "Relays should limit the use of `r` query params on the redirct url to up to 2 to avoid infinite redirect loop"

Reply:
I think it's better if you just let clients query each host, relays can simply 404 and clients will try the next relay

Reply:
Hello @v0l thanks for reviewing. Allow me to try to convince you.

The idea is taking into account how urls are used on kind 1 notes, DMs and long-form content notes. Like embedded in a `.content` as the string `"Nice sunset! \n\n https://relay.domain/nip95/<sha256-file-hash>?r=relay.domain2&r=relay.domain3" \n\n Thanks for coming.`. (Side note, we should add extension like .jpg to help clients detect it as image for instance)

If it were an embedded `nostr:nevent1...` url that points to an event with extra relay tags (or extra url tags), it could work, but I'm 10000% sure Damus wouldn't use it cause it would need another hop to download the event (@jb55 [doesn't like embedding nip-94 for instance](https://github.com/nostr-protocol/nips/pull/478#issuecomment-1526142478)).

So a media reource on a note using the url like the above would live longer and could simply be added to an \<img\> or \<video\> tag which would be enough to try downloading from 1 to 3 relays. Makes sense or still no deal?

Reply:
I just think its a bit dangerous to follow redirects from random relays. It also seems a bit unnecessary, relays might just not implement the redirect response and return 404 and then we're back to just doing it in clients anyway.

Having the redirect issued by the server also makes it difficult to implement this NIP via a dumb nginx host.

Reply:
> its a bit dangerous to follow redirects from random relays

The server won't follow the redirect when the server replies with a 302 status, the client will. I'm not asking the server to act as a proxy, just redirect. It would be the "same" as client trying manually as you said, when receiving a 404.

> [...] (it) makes it difficult to implement this NIP via a dumb nginx host

I will make it a "SHOULD" instead of "MUST" on the NIP text ok? Not required.

Reply:
IMO it needs to be MUST or remove it from the spec, get a few more opinions on the redirects, my vote would be to remove it

Reply:
Yea i know why you want to use the `r` it makes sense, but this will result in redirects to random images from bad relays, you need to be checking the hash of the image and so you need to intercept the response no matter what.

Reply:
@v0l I just removed the `r` tag because people wanted to redirect to CDNs so they wouldn't be able to configure dynamic redirects from CDNs to other servers easily in case of missing file.